config/configtls/configtls.go

```diff
@@ -85,6 +91,59 @@ type TLSServerSetting struct {
 	ClientCAFile string `mapstructure:"client_ca_file"`
 }
 
+// certReloader is a wrapper object for certificate reloading
+// Its GetCertificate method will either return the current certificate or reload from disk
+// if the last reload happened more than ReloadInterval ago
+type certReloader struct {
+	// Path to the TLS cert
+	CertFile string
+	// Path to the TLS key
+	KeyFile string
+	// ReloadInterval specifies the duration after which the certificate will be reloaded
+	// If not set, it will never be reloaded (optional)
+	ReloadInterval time.Duration
+	lastReload     time.Time
+	cert           *tls.Certificate
+	lock           sync.RWMutex
+}
+
+func newCertReloader(certFile, keyFile string, reloadInterval time.Duration) (*certReloader, error) {
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+	return &certReloader{
+		CertFile:       certFile,
+		KeyFile:        keyFile,
+		ReloadInterval: reloadInterval,
+		lastReload:     time.Now(),
+		cert:           &cert,
+	}, nil
+}
+
+func (r *certReloader) GetCertificate() (*tls.Certificate, error) {
+	now := time.Now()
+	// Read locking here before we do the time comparison
+	// If a reload is in progress this will block and we will skip reloading in the current
+	// call once we can continue
+	r.lock.RLock()
+	if r.ReloadInterval != 0 && r.lastReload.Add(r.ReloadInterval).Before(now) {
+		// Need to release the read lock, otherwise we deadlock
+		r.lock.RUnlock()
+		r.lock.Lock()
+		defer r.lock.Unlock()
+		cert, err := tls.LoadX509KeyPair(r.CertFile, r.KeyFile)
+		if err != nil {
+			return nil, fmt.Errorf("failed to load TLS cert and key: %w", err)
+		}
+		r.cert = &cert
+		r.lastReload = now
+		return r.cert, nil
+	}
+	defer r.lock.RUnlock()
```

Comment:
If you really care about the extra cycle, you can:
1. Precalculate the "next" reload instead of always calculating `r.lastReload.Add(r.ReloadInterval)`
2. Can use an atomic variable for the "nextReloadTime" instead of a rlock, that will complicate a bit the logic, but can be done.

Reply:
I would for the moment drop the 2. point, but first may worth doing.

Reply:
precalculating sounds good, I'll add that in.

I'm not super sold on the atomic variable: I don't think this code is that performance critical to justify it.